source/extensions/quic_listeners/quiche/envoy_quic_server_stream.cc

```diff
@@ -31,15 +31,23 @@ EnvoyQuicServerStream::EnvoyQuicServerStream(quic::QuicStreamId id, quic::QuicSp
                                              quic::StreamType type)
     : quic::QuicSpdyServerStreamBase(id, session, type),
       EnvoyQuicStream(
-          2 * GetQuicFlag(FLAGS_quic_buffered_data_threshold),
-          [this]() { runLowWatermarkCallbacks(); }, [this]() { runHighWatermarkCallbacks(); }) {}
+          // This should be larger than 8k to fully utilize congestion control
+          // window. And no larger than the max stream flow control window for
+          // the stream to buffer all the data.
+          // Ideally this limit should also corelate to peer's receive window
+          // but not fully depends on that.
+          16 * 1024, [this]() { runLowWatermarkCallbacks(); },
```

Comment:
So basically if this is what we're willing to buffer shipping data downstream, we need it to be as large as the amount of data we can get from upstream.
I don't think there's a way from the HCM to get the upstream H2 codec limit. I'd suggest defaulting it to 2x the DEFAULT_INITIAL_STREAM_WINDOW_SIZE with a TODO that when we assign an upstream connection we need to plumb through the limits based on config.

Reply:
@alyssawilk do you have any thoughts on regarding my follow up comment?

Reply:
See my comment above. IMO for QUIC we should start with good edge defaults, since this is really practically only going to be used at the edge. So yeah 256MiB default is way too large. Can we just use different defaults for QUIC and get them right now? If we want to TODO this for discussion during a production readiness pass that is fine with me.

Reply:
16K here is the soft limit for quic stream send buffer. It doesn't have to be associated with upstream flow control window. Basically quic can tell upstream to stop shipping data whenever it desires.  If you think this number is okay than that's it.  To fully utilize congestion control window, it's better to be larger than 8K.

Reply:
Yeah, I thought this one was tied to the upstream limit, i.e. if we set it low we'd end up advertising to the peer we were willing to buffer data when we wouldn't locally.  Sounds like I had the wrong field, they can be decoupled and we can start with smaller limits, which SGTM.